src/test/java/org/elasticsearch/gateway/GatewayMetaStateTests.java

```diff
@@ -0,0 +1,179 @@
+        MetaData metaDataNewClusterState = MetaData.builder()
+                .put(init.metaData().index("test"), versionChanged)
+                .build();
+
+        // create the cluster states with meta data and routing tables as computed before
+        previousClusterState = ClusterState.builder(init)
+                .metaData(metaDataOldClusterState)
+                .routingTable(routingTableOldClusterState)
+                .nodes(DiscoveryNodes.builder().put(newNode("node1")).localNodeId("node1"))
+                .build();
+        newClusterState = ClusterState.builder(previousClusterState).routingTable(routingTableNewClusterState).metaData(metaDataNewClusterState).build();
+
+        return new ClusterChangedEvent("test", newClusterState, previousClusterState);
+    }
+
+    public void assertState(boolean initializing,
+                            boolean versionChanged,
+                            boolean stateOnDisk,
+                            boolean stateOnDiskOutdated,
+                            boolean stateInMemory,
+                            boolean stateInMemoryOutdated,
+                            boolean masterEligible,
+                            boolean expectMetaData) throws Exception {
+        ClusterChangedEvent event = generateEvent(initializing, versionChanged);
+        final NodeEnvironment env = newNodeEnvironment();
+        MetaStateService metaStateService = new MetaStateService(ImmutableSettings.EMPTY, env);
+        IndexMetaState indexMetaState = new IndexMetaState(ImmutableSettings.EMPTY, env, metaStateService);
+
+        MetaData inMemoryMetaData = null;
+        if (stateInMemory) {
+            if (stateInMemoryOutdated) {
+                inMemoryMetaData = event.previousState().metaData();
+            } else {
+                inMemoryMetaData = event.state().metaData();
+            }
+        }
+        if (stateOnDisk) {
+            if (stateOnDiskOutdated) {
+                metaStateService.writeIndex("test", event.previousState().metaData().index("test"), null);
+            } else {
+                metaStateService.writeIndex("test", event.state().metaData().index("test"), null);
+            }
+        }
+        Iterator<GatewayMetaState.IndexMetaWriteInfo> indices;
+        if (masterEligible) {
+            indices = indexMetaState.getIndicesToWriteMasterNode(event, inMemoryMetaData).iterator();
+        } else {
+            indices = indexMetaState.getIndicesToWriteDataOnlyNode(event, inMemoryMetaData).iterator();
+        }
+        if (expectMetaData) {
+            assertThat(indices.hasNext(), equalTo(true));
+            assertThat(indices.next().getNewMetaData().index(), equalTo("test"));
+            assertThat(indices.hasNext(), equalTo(false));
+        } else {
+            assertThat(indices.hasNext(), equalTo(false));
+        }
+    }
+
+    @Test
+    public void testAllCombinations() throws Exception {
```

Comment:
can we have a test for each of these names after what they are doing? that would be very helpful I think! I am ok with duplicating code here and there but these unittests should be more readable